src/shared-application.cpp

```diff
@@ -0,0 +1,180 @@
+        event_base_loopbreak(ev_base);
+    }
+}
+
+#ifndef Q_OS_WIN32
+void *askActivateSynchronically(void * /*arg*/) {
+#else
+DWORD WINAPI askActivateSynchronically(LPVOID /*arg*/) {
+#endif
+    _CcnetClient* async_client = ccnet_client_new();
+    _CcnetClient *sync_client = ccnet_client_new();
+    const QString ccnet_dir = defaultCcnetDir();
+    if (ccnet_client_load_confdir(async_client, toCStr(ccnet_dir)) <  0) {
+        g_object_unref(sync_client);
+        g_object_unref(async_client);
+        return 0;
+    }
+
+    if (ccnet_client_connect_daemon(async_client, CCNET_CLIENT_ASYNC) < 0) {
+        g_object_unref(sync_client);
+        g_object_unref(async_client);
+        return 0;
+    }
+    if (ccnet_client_load_confdir(sync_client, toCStr(ccnet_dir)) <  0) {
+        g_object_unref(sync_client);
+        g_object_unref(async_client);
+        return 0;
+    }
+
+    if (ccnet_client_connect_daemon(sync_client, CCNET_CLIENT_SYNC) < 0) {
+        g_object_unref(sync_client);
+        g_object_unref(async_client);
+        return 0;
+    }
+    //
+    // send message synchronously
+    //
+    CcnetMessage *syn_message;
+    syn_message = ccnet_message_new(sync_client->base.id,
+                                    sync_client->base.id,
+                                    kAppletCommandsMQ, "syn_activate", 0);
+
+    // blocking io, but cancellable from pthread_cancel
+    if (ccnet_client_send_message(sync_client, syn_message) < 0) {
+        ccnet_message_free(syn_message);
+        g_object_unref(sync_client);
+        g_object_unref(async_client);
+        return 0;
+    }
+
+    ccnet_message_free(syn_message);
+
+    //
+    // receive message asynchronously
+    //
+    struct event_base* ev_base = event_base_new();
+    struct timeval timeout = {
+        .tv_sec = 1,
+        .tv_usec = 0
+    };
+    // set timeout
+    event_base_loopexit(ev_base, &timeout);
+
+    UserData user_data;
+    user_data.ev_base = ev_base;
+    user_data.ccnet_client = async_client;
+    struct event *read_event = event_new(ev_base, async_client->connfd,
+                                         EV_READ | EV_PERSIST, readCallback, &user_data);
+    event_add(read_event, NULL);
+
+    // set message read callback
+    _CcnetMqclientProc *mqclient_proc = (CcnetMqclientProc *)
+        ccnet_proc_factory_create_master_processor
+        (async_client->proc_factory, "mq-client");
+
+    ccnet_mqclient_proc_set_message_got_cb(mqclient_proc,
+                                           messageCallback,
+                                           ev_base);
+
+    const char *topics[] = {
+        kAppletCommandsMQ,
+    };
+
+    if (ccnet_processor_start((CcnetProcessor *)mqclient_proc,
+                              G_N_ELEMENTS(topics), (char **)topics) < 0) {
+        event_base_free(ev_base);
+        g_object_unref(sync_client);
+        g_object_unref(async_client);
+        return 0;
+    }
+
+    event_base_dispatch(ev_base);
+
```

Comment:
Directly handle libevent doesn't seem so good to me. Maybe we could provide a rpc service in applet, and let the newly started applet instance invoke the rpc, instead of using ccnet messages? (messages are more suitable for broadcasting, or pub/sub).

The